File: 02-addTwoNumbers/addTwoNumbers.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    dummy := &ListNode{} //哨兵节点
    curr := dummy
    carry := 0
    for l1 != nil || l2 != nil || carry != 0 {
        sum := 0
        if carry != 0 {
            sum += carry
        }
        if l1 != nil {
            sum += l1.Val
        } 
        if l2 != nil {
            sum += l2.Val
        }
        if sum >= 10 {
            carry = 1
            sum -= 10
        } else {
            carry = 0
        }
        curr.Next = &ListNode{sum, nil}
        curr = curr.Next
        if l1 != nil {
            l1 = l1.Next
        }
        if l2 != nil {
            l2 = l2.Next
        }
    }
    return dummy.Next
}

File: 02-addTwoNumbers/addTwoNumbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *dummyNode = new ListNode;
        ListNode *curr = dummyNode;
        int carry = 0;
        while (l1 || l2 || carry != 0) {
            int sum = carry;
            if (l1) {
                sum += l1->val;
            }
            if (l2) {
                sum += l2->val;
            }
            if (sum >= 10) {
                carry = 1;
                sum -= 10;
            } else {
                carry = 0;
            }
            ListNode *node = new ListNode{sum, nullptr};
            curr->next = node;
            curr = curr->next;
            if (l1) {
                l1 = l1->next;
            } 
            if (l2) {
                l2 = l2->next;
            }
        }
        return dummyNode->next;
    }
};
